### ISSUE TITLE
[Bug] Upgrade ember-source to 3.25.x breaks on compiling templates

### ISSUE THREAD

**Issue by scottmessinger** [emberjs/ember.js#19482]
### 🐞 Describe the Bug
After upgrading to ember-source 3.25.X, I get the following error when doing `ember serve`

```
Template Compiler Error (TemplateCompiler) in cc-frontend/templates/-user-after-upgrade.hbs

Cannot read property 'start' of null


Stack Trace and Error Report: /var/folders/tn/vw59jr41329bxq_n5nz8ws8m0000gn/T/error.dump.f5a8520a939e7644822798a8292331c6.log
```

<details>
<summary>Stack trace</summary>

```sh
=================================================================================

ENV Summary:

  TIME: Fri Mar 26 2021 14:49:19 GMT-0400 (Eastern Daylight Time)
  TITLE: ember
  ARGV:
  - /Users/scottamesmessinger/.volta/tools/image/node/10.15.3/bin/node
  - /Users/scottamesmessinger/code/commoncurriculum/v4/app/frontend/node_modules/.bin/ember
  - serve
  - -p
  - 4200
  EXEC_PATH: /Users/scottamesmessinger/.volta/tools/image/node/10.15.3/bin/node
  TMPDIR: /var/folders/tn/vw59jr41329bxq_n5nz8ws8m0000gn/T
  SHELL: /bin/zsh
  PATH:
  - /Users/scottamesmessinger/.volta/tools/image/node/10.15.3/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin
  - /Users/scottamesmessinger/code/commoncurriculum/v4/app/frontend/node_modules/.bin
  - /Users/scottamesmessinger/.volta/tools/image/yarn/1.19.1/bin
  - /Users/scottamesmessinger/.volta/tools/image/node/10.15.3/bin
  - /Users/scottamesmessinger/.asdf/shims
  - /opt/homebrew/opt/asdf/bin
  - /Users/scottamesmessinger/.volta/bin
  - /opt/homebrew/bin
  - /opt/homebrew/sbin
  - /usr/local/bin
  - /usr/bin
  - /bin
  - /usr/sbin
  - /sbin
  PLATFORM: darwin x64
  FREEMEM: 66916352
  TOTALMEM: 8589934592
  UPTIME: 2926
  LOADAVG: 4.42333984375,4.189453125,4.93310546875
  CPUS:
  - Apple M1 - 2400
  - Apple M1 - 2400
  - Apple M1 - 2400
  - Apple M1 - 2400
  - Apple M1 - 2400
  - Apple M1 - 2400
  - Apple M1 - 2400
  - Apple M1 - 2400
  ENDIANNESS: LE
  VERSIONS:
  - ares: 1.15.0
  - cldr: 33.1
  - http_parser: 2.8.0
  - icu: 62.1
  - modules: 64
  - napi: 3
  - nghttp2: 1.34.0
  - node: 10.15.3
  - openssl: 1.1.0j
  - tz: 2018e
  - unicode: 11.0
  - uv: 1.23.2
  - v8: 6.8.275.32-node.51
  - zlib: 1.2.11

ERROR Summary:

  - broccoliBuilderErrorStack: TypeError: Cannot read property 'start' of null
    at Source.spanFor (/Users/scottamesmessinger/code/commoncurriculum/v4/app/frontend/node_modules/ember-source/dist/ember-template-compiler.js:9071:26)
    at BlockContext.loc (/Users/scottamesmessinger/code/commoncurriculum/v4/app/frontend/node_modules/ember-source/dist/ember-template-compiler.js:14161:26)
    at StatementNormalizer.BlockStatement (/Users/scottamesmessinger/code/commoncurriculum/v4/app/frontend/node_modules/ember-source/dist/ember-template-compiler.js:14484:31)
    at StatementNormalizer.normalize (/Users/scottamesmessinger/code/commoncurriculum/v4/app/frontend/node_modules/ember-source/dist/ember-template-compiler.js:14402:23)
    at /Users/scottamesmessinger/code/commoncurriculum/v4/app/frontend/node_modules/ember-source/dist/ember-template-compiler.js:14572:27
    at Array.map (<anonymous>)
    at ElementNormalizer.ElementNode (/Users/scottamesmessinger/code/commoncurriculum/v4/app/frontend/node_modules/ember-source/dist/ember-template-compiler.js:14571:41)
    at StatementNormalizer.normalize (/Users/scottamesmessinger/code/commoncurriculum/v4/app/frontend/node_modules/ember-source/dist/ember-template-compiler.js:14405:52)
    at /Users/scottamesmessinger/code/commoncurriculum/v4/app/frontend/node_modules/ember-source/dist/ember-template-compiler.js:14572:27
    at Array.map (<anonymous>)
  - code: [undefined]
  - codeFrame: Cannot read property 'start' of null
  - errorMessage: cc-frontend/templates/-user-after-upgrade.hbs: Cannot read property 'start' of null
        in /var/folders/tn/vw59jr41329bxq_n5nz8ws8m0000gn/T/broccoli-5228kgR8ffkjEBMP/out-913-colocated_template_processor
        at TemplateCompiler
  - errorType: Template Compiler Error
  - location:
    - column: [undefined]
    - file: cc-frontend/templates/-user-after-upgrade.hbs
    - line: [undefined]
    - treeDir: /var/folders/tn/vw59jr41329bxq_n5nz8ws8m0000gn/T/broccoli-5228kgR8ffkjEBMP/out-913-colocated_template_processor
  - message: cc-frontend/templates/-user-after-upgrade.hbs: Cannot read property 'start' of null
        in /var/folders/tn/vw59jr41329bxq_n5nz8ws8m0000gn/T/broccoli-5228kgR8ffkjEBMP/out-913-colocated_template_processor
        at TemplateCompiler
  - name: Error
  - nodeAnnotation: [undefined]
  - nodeName: TemplateCompiler
  - originalErrorMessage: Cannot read property 'start' of null
  - stack: TypeError: Cannot read property 'start' of null
    at Source.spanFor (/Users/scottamesmessinger/code/commoncurriculum/v4/app/frontend/node_modules/ember-source/dist/ember-template-compiler.js:9071:26)
    at BlockContext.loc (/Users/scottamesmessinger/code/commoncurriculum/v4/app/frontend/node_modules/ember-source/dist/ember-template-compiler.js:14161:26)
    at StatementNormalizer.BlockStatement (/Users/scottamesmessinger/code/commoncurriculum/v4/app/frontend/node_modules/ember-source/dist/ember-template-compiler.js:14484:31)
    at StatementNormalizer.normalize (/Users/scottamesmessinger/code/commoncurriculum/v4/app/frontend/node_modules/ember-source/dist/ember-template-compiler.js:14402:23)
    at /Users/scottamesmessinger/code/commoncurriculum/v4/app/frontend/node_modules/ember-source/dist/ember-template-compiler.js:14572:27
    at Array.map (<anonymous>)
    at ElementNormalizer.ElementNode (/Users/scottamesmessinger/code/commoncurriculum/v4/app/frontend/node_modules/ember-source/dist/ember-template-compiler.js:14571:41)
    at StatementNormalizer.normalize (/Users/scottamesmessinger/code/commoncurriculum/v4/app/frontend/node_modules/ember-source/dist/ember-template-compiler.js:14405:52)
    at /Users/scottamesmessinger/code/commoncurriculum/v4/app/frontend/node_modules/ember-source/dist/ember-template-compiler.js:14572:27
    at Array.map (<anonymous>)

=================================================================================
```

</details>

The offending template looks like this: 

```hbs
<div class="user-after-upgrade__text">
  <div class="generic-contentful-wrapper">
    <ContentfulData @contentfulEntryId="5apr2UOGveiIo20cu4oG4k" as |entry|>
      {{commonmark-it entry.attributes.text}}
    </ContentfulData>
  </div>
</div>
```

If I remove that code, I get the same error on a different template that looks quite different: 

```hbs
<Admin::Skeleton
  @model={{model}}
  @sidebarComponent="admin/user/sidebar"
  @detailComponent="admin/user/detail"
 />
```

I can't tell any aspects of the template that would be problematic.

### 🔬 Minimal Reproduction
I assume this is mismatched dependency -- if someone can point me in the right direction, I can try to narrow it down. There are so many dependencies that seem to touch template compilation, I'm a bit at a loss.

### 😕 Actual Behavior
Should build without a problem.

### 🤔 Expected Behavior
Broken build

### 🌍 Environment

- Ember: - 3.25.4 (though breaks on all 3.25.X versions)
- Node.js/npm: - node@10.15.3 and yarn@1.19.1 and npm@6.9.0
- OS: - MacOS 11.2 on Apple Silicon
- Browser: - Safari 14.0.3

### ➕ Additional Context

<details>
<summary>package.json</summary>

```js
{
  "devDependencies": {
    "@ember/optional-features": "^1.3.0",
    "@ember/render-modifiers": "^1.0.2",
    "@ember/test-helpers": "2.2.0",
    "@glimmer/component": "^1.0.3",
    "@glimmer/tracking": "^1.0.3",
    "@storybook/addon-actions": "^5.3.18",
    "@storybook/addon-links": "^5.3.18",
    "@storybook/addons": "^5.3.18",
    "@storybook/ember": "^5.3.18",
    "@types/ember": "^3.16.2",
    "@types/ember-qunit": "^3.4.13",
    "@types/ember-resolver": "^5.0.10",
    "@types/ember-testing-helpers": "0.0.4",
    "@types/ember__object": "^3.1.3",
    "@types/ember__service": "^3.0.6",
    "@types/ember__test-helpers": "^1.7.3",
    "@types/lodash-es": "^4.17.3",
    "@types/qunit": "^2.11.0",
    "@types/rsvp": "^4.0.3",
    "@typescript-eslint/eslint-plugin": "^4.14.1",
    "@typescript-eslint/parser": "^4.14.1",
    "algoliasearch": "4.5.1",
    "babel-eslint": "^10.1.0",
    "babel-loader": "^8.1.0",
    "broccoli-asset-rev": "^3.0.0",
    "ember-animated": "^0.10.1",
    "ember-animated-tools": "^0.4.1",
    "ember-ast-hot-load": "^0.1.2",
    "ember-async-await-helper": "^0.2.0",
    "ember-auto-import": "^1.11.2",
    "ember-autoresize-modifier": "^0.3.0",
    "ember-body-class": "^1.1.4",
    "ember-changeset": "3.3.4",
    "ember-changeset-validations": "3.3.1",
    "ember-cli": "^3.25.2",
    "ember-cli-app-version": "^5.0.0",
    "ember-cli-babel": "^7.23.1",
    "ember-cli-bundle-analyzer": "^0.2.1",
    "ember-cli-cjs-transform": "^2.0.0",
    "ember-cli-clipboard": "^0.14.0",
    "ember-cli-custom-assertions": "0.2.0",
    "ember-cli-dependency-checker": "^3.2.0",
    "ember-cli-dependency-lint": "^1.1.3",
    "ember-cli-deploy": "^1.0.2",
    "ember-cli-deploy-build": "^2.0.0",
    "ember-cli-deploy-display-revisions": "^1.0.1",
    "ember-cli-deploy-fastboot-app-server": "^0.0.3",
    "ember-cli-deploy-fastboot-app-server-aws": "^0.0.4",
    "ember-cli-deploy-manifest": "^1.1.0",
    "ember-cli-deploy-revision-data": "commoncurriculum/ember-cli-deploy-revision-data",
    "ember-cli-deploy-s3": "^1.4.0",
    "ember-cli-eslint": "^5.1.0",
    "ember-cli-fastboot": "2.2.2",
    "ember-cli-get-component-path-option": "^1.0.0",
    "ember-cli-head": "^0.4.1",
    "ember-cli-htmlbars": "^5.7.1",
    "ember-cli-inject-live-reload": "^2.0.2",
    "ember-cli-inline-content": "^0.4.1",
    "ember-cli-is-package-missing": "^1.0.0",
    "ember-cli-moment-shim": "3.7.1",
    "ember-cli-normalize-entity-name": "^1.0.0",
    "ember-cli-postcss": "^6.0.1",
    "ember-cli-release": "^0.2.9",
    "ember-cli-sass": "^10.0.1",
    "ember-cli-sauce": "2.2.0",
    "ember-cli-shims": "^1.2.0",
    "ember-cli-sri": "^2.1.1",
    "ember-cli-string-helpers": "^4.0.6",
    "ember-cli-string-utils": "^1.1.0",
    "ember-cli-template-lint": "^2.0.2",
    "ember-cli-terser": "^4.0.0",
    "ember-cli-test-info": "^1.0.0",
    "ember-cli-test-loader": "3.0.0",
    "ember-cli-typescript": "^4.0.0",
    "ember-cli-typescript-blueprints": "^3.0.0",
    "ember-cli-update": "^0.54.6",
    "ember-cli-valid-component-name": "^1.0.0",
    "ember-cli-version-checker": "^5.1.1",
    "ember-composable-helpers": "4.4.1",
    "ember-concurrency": "1.1.7",
    "ember-concurrency-decorators": "1.0.0",
    "ember-concurrency-scroll": "1.0.0",
    "ember-concurrency-ts": "^0.2.0",
    "ember-cookies": "^0.5.2",
    "ember-could-get-used-to-this": "^1.0.1",
    "ember-css-modules": "1.3.0",
    "ember-decorators": "^6.1.1",
    "ember-elsewhere": "1.1.0",
    "ember-exam": "^6.0.1",
    "ember-export-application-global": "^2.0.1",
    "ember-fetch": "^8.0.4",
    "ember-in-viewport": "^3.7.3",
    "ember-keyboard": "^5.0.0",
    "ember-load-initializers": "^2.1.1",
    "ember-lodash": "^4.19.5",
    "ember-math-helpers": "^2.14.0",
    "ember-maybe-import-regenerator": "^0.1.6",
    "ember-modal-dialog": "3.0.0-beta.4",
    "ember-moment": "7.8.1",
    "ember-notifier": "0.1.3",
    "ember-page-title": "^6.2.1",
    "ember-parachute": "^1.0.2",
    "ember-power-calendar": "^0.15.0",
    "ember-power-calendar-moment": "0.1.7",
    "ember-power-select": "^4.1.0",
    "ember-power-select-with-create": "github:commoncurriculum/ember-power-select-with-create#dead7bbaaa11cb1dad4db687692ef72381a5fd83",
    "ember-query-params-service": "0.5.5",
    "ember-qunit": "^5.1.2",
    "ember-resolver": "^8.0.0",
    "ember-responsive": "^3.0.6",
    "ember-set-helper": "^2.0.0",
    "ember-sortable": "git://github.com/commoncurriculum/ember-sortable.git#65743e6189b4a24133b181d450b31363e5b3cf92",
    "ember-source": "3.25.4",
    "ember-stargate": "^0.2.0",
    "ember-statecharts": "^0.13.2",
    "ember-table": "2.2.3",
    "ember-template-lint": "^2.18.1",
    "ember-template-lint-plugin-prettier": "^1.1.0",
    "ember-test-helpers-codemod": "^0.2.0",
    "ember-test-selectors": "^4.1.0",
    "ember-tether": "^2.0.0",
    "ember-toggle": "github:commoncurriculum/ember-toggle#e3f4704b4a88dc5dade96c3fc58289c1e3d06ccb",
    "ember-truth-helpers": "^2.1.0",
    "ember-usable": "https://github.com/pzuraq/ember-usable#0d03a50",
    "ember-welcome-page": "^4.0.0",
    "eslint": "^7.20.0",
    "eslint-config-prettier": "^7.2.0",
    "eslint-plugin-ember": "^10.2.0",
    "eslint-plugin-hbs": "^1.0.0",
    "eslint-plugin-node": "^11.1.0",
    "eslint-plugin-prettier": "^3.3.1",
    "fastboot": "^3.0.3",
    "fastboot-app-server": "2.0.0",
    "fastboot-express-middleware": "2.0.0",
    "js-search": "^2.0.0",
    "liquid-fire": "0.31.0",
    "liquid-tether": "^2.0.7",
    "liquid-wormhole": "^2.1.5",
    "loader.js": "^4.7.0",
    "lodash-es": "^4.17.15",
    "morgan": "^1.10.0",
    "postcss-clean": "^1.1.0",
    "prettier": "^2.2.1",
    "qunit": "^2.14.0",
    "qunit-dom": "1.6.0",
    "sass": "^1.26.5",
    "tailwindcss": "0.7.4",
    "testem": "^3.1.0",
    "throng": "4.0.0",
    "tracked-built-ins": "^1.0.2",
    "tracked-toolbox": "github:commoncurriculum/tracked-toolbox#5c067d65a19ee16f3188a9efba6c9f6ea378d73b",
    "typescript": "^4.1.2",
    "webpack": "^4.43.0"
  },
  "dependencies": {
    "@emberwatch/ember-language-server": "^0.2.1",
    "@marcj/marshal": "^2.1.6",
    "@sentry/browser": "^6.1.0",
    "@sentry/node": "^6.1.0",
    "@storybook/addon-centered": "^5.3.18",
    "@storybook/addon-knobs": "^5.3.18",
    "@storybook/addon-notes": "^5.3.18",
    "@storybook/addon-storysource": "^5.3.18",
    "@storybook/ember-cli-storybook": "storybookjs/ember-cli-storybook#master",
    "@types/node": "^13.9.0",
    "autoprefixer": "^9.7.6",
    "broccoli-file-creator": "^2.1.1",
    "broccoli-filter": "^1.3.0",
    "broccoli-multi-filter": "^1.2.3-multi.3",
    "broccoli-multifilter": "^2.0.0",
    "browserslist": "^4.12.0",
    "compare-versions": "^3.6.0",
    "ember-intersection-observer-modifier": "^1.0.0",
    "ember-raf-scheduler": "^0.1.0",
    "fastboot": "^3.0.3",
    "fastboot-pool": "^0.6.3",
    "immer": "^6.0.5",
    "js-base64": "^2.5.2",
    "js-cookie": "^2.2.1",
    "moment-range": "^4.0.2",
    "morgan": "^1.10.0",
    "node-fetch": "^2.6.0",
    "node-sass": "^4.14.1",
    "pm2": "^4.4.0",
    "postcss-nested": "^4.2.1",
    "prismic-dom": "^2.2.3",
    "reflect-metadata": "^0.1.13",
    "ua-parser-js": "*",
    "winston": "2.3.1"
  },
  "fastbootDependencies": [
    "ua-parser-js",
    "moment-range",
    "winston"
  ],
  "optionalDependencies": {
    "fsevents": "^2.1.3"
  },
  "volta": {
    "node": "10.15.3",
    "yarn": "1.19.1"
  },
  "browserslist": [
    "last 3 Edge versions",
    "last 3 Chrome versions",
    "last 3 Firefox versions",
    "last 3 Safari versions",
    "> 2%"
  ]
}
```

</details>

**Comment by pzuraq** [comment]
So it's hard to tell for sure, but the biggest reason we've seen issues like this is actually because of custom Handlebars template transforms which were _not_ using the builder API. These transforms would leave out crucial node information, like `loc` info, which resulted in malformed nodes. The Handlebars AST and transforms are not something that we consider public API, so this failure would be expected if that was the case.

Are there any addons you're using that could be doing template preprocessing? If you're unsure, I can try to help figure out a way to give you list of all transforms that are running in the template compiler and we can start there.

**Comment by scottmessinger** [comment]
@pzuraq Thanks! I'm using ember-css-modules which I think does some transforms. Is there a way I find this out so as not to bother you with listing them out? Or, is this something that isn't really discoverable by end users?

**Comment by pzuraq** [comment]
It looks like `ember-css-modules` uses the builder system properly, so I don't think it's that one 🤔  also it appears to have passed tests on 3.25+ I think.

For listing transforms, you have to dig into `ember-template-compiler` in `node_modules`, it's a bit of a pain. You'll want to put a `console.log` around here: https://github.com/emberjs/ember.js/blob/master/packages/ember-template-compiler/lib/system/compile-options.ts#L64

I _believe_ the plugins have `.name` properties that can help you figure out where they're from usually, but some may not. That's the starting point though, sorry about the complexity here 😕  it would be good to make a better debugging API

**Comment by kategengler** [comment]
I believe older versions of `ember-basic-dropdown` are one of the culprits for the error you describe https://github.com/cibernox/ember-basic-dropdown/issues/586 It's a dependency of the `ember-power-*` addons.

**Comment by scottmessinger** [comment]
@kategengler Thanks so much! I'll look at that. I'm coming up empty in figuring out which transform is breaking things.

**Comment by scottmessinger** [comment]
@pzuraq When I add a console.log line where you said, nothing ever hits that line. If I log the `options` returned at the end of that method, I get these: 

```
[Function: transformOldClassBindingSyntax],
        [Function: transformQuotedBindingsIntoJustBindings],
        [Function: assertReservedNamedArguments],
        [Function: transformActionSyntax],
        [Function: transformAttrsIntoArgs],
        [Function: transformEachInIntoEach],
        [Function: transformHasBlockSyntax],
        [Function: transformLinkTo],
        [Function: errorOnInputWithContent],
        [Function: transformInElement],
        [Function: assertSplattributeExpressions],
        [Function: transformEachTrackArray],
        [Function: transformWrapMountAndOutlet],
        [Function: deprecateSendAction],
        [Function: assertAgainstDynamicHelpersModifiers]
```

Is that the plugins you're asking about? And, should I then look for the addons that provide those transforms?

**Comment by scottmessinger** [comment]
@pzuraq While I'm not seeing the branch you highlighted ever being reached, I can log the USER_PLUGINS and builtInPlugins.

The only USER_PLUGINS that sometimes comes up is: `StripTestSelectorsTransform`

The `builtInPlugins` are:
```
  [Function: transformOldClassBindingSyntax],
  [Function: transformQuotedBindingsIntoJustBindings],
  [Function: assertReservedNamedArguments],
  [Function: transformActionSyntax],
  [Function: transformAttrsIntoArgs],
  [Function: transformEachInIntoEach],
  [Function: transformHasBlockSyntax],
  [Function: transformLinkTo],
  [Function: errorOnInputWithContent],
  [Function: transformInElement],
  [Function: assertSplattributeExpressions],
  [Function: transformEachTrackArray],
  [Function: transformWrapMountAndOutlet],
  [Function: deprecateSendAction],
  [Function: assertAgainstDynamicHelpersModifiers]
```

**Comment by scottmessinger** [comment]
I've been trying to log the block/ast and the log lines right above the line where it errors are never being logged to the console. I can't figure out what's going on. I tried `CLEAR_BROCCOLI_PERSISTENT_FILTER_CACHE=true ember build` but that didn't help. 

I'm really confused why I can't even get log messages where the error is happening.

**Comment by scottmessinger** [comment]
I still have no idea why the logs are being swallowed, but I did figure out which add-on is making the build break: https://github.com/lifeart/ember-ast-hot-load I have no idea why -- I'm on the latest version (1.2.3) and it seems up to date. Perhaps it's conflicting with another add-on? 

Either way, thanks @pzuraq for the help. Working from your idea, I ended up going into the Broccoli lib to add more logs and found my way into the ember template colocation plugin and eventually, was able to log closer to the point of the error. Somewhere in there, I removed `ember-ast-hot-load` and things started working. Your tip helped start me on a debugging journey that helped me make progress. Thanks!
